node/network/collator-protocol/src/peer_slots.rs

```diff
@@ -0,0 +1,639 @@
+
+use rand::{thread_rng, Rng, rngs::ThreadRng};
+use rand::prelude::SliceRandom;
+
+use sp_keystore::SyncCryptoStorePtr;
+
+use polkadot_node_network_protocol::{UnifiedReputationChange as Rep, View};
+use polkadot_primitives::v1::{Hash, Id as ParaId};
+use polkadot_subsystem::{messages::{AllMessages, NetworkBridgeMessage}, SubsystemSender};
+
+pub(crate) const CACHE_SIZE: usize = 10_000;
+pub(crate) const RESERVOIR_SIZE: usize = 1_000;
+
+pub(crate) const COST_UNEXPECTED_MESSAGE: Rep = Rep::CostMinor("An unexpected message");
+/// Message could not be decoded properly.
+pub(crate) const COST_CORRUPTED_MESSAGE: Rep = Rep::CostMinor("Message was corrupt");
+/// Network errors that originated at the remote host should have same cost as timeout.
+pub(crate) const COST_NETWORK_ERROR: Rep = Rep::CostMinor("Some network error");
+pub(crate) const COST_REQUEST_TIMED_OUT: Rep =
+	Rep::CostMinor("A collation request has timed out");
+pub(crate) const COST_INVALID_SIGNATURE: Rep =
+	Rep::Malicious("Invalid network message signature");
+pub(crate) const COST_WRONG_PARA: Rep =
+	Rep::Malicious("A collator provided a collation for the wrong para");
+pub(crate) const COST_UNNEEDED_COLLATOR: Rep = Rep::CostMinor("An unneeded collator connected");
+
+pub(crate) const COST_REPORT_BAD: Rep =
+	Rep::Malicious("A collator was reported by another subsystem");
+pub(crate) const BENEFIT_NOTIFY_GOOD: Rep =
+	Rep::BenefitMinor("A collator was noted good by another subsystem");
+
+const COLLATOR_METRICS: [Rep; 9] = [
+	COST_UNEXPECTED_MESSAGE,
+	COST_CORRUPTED_MESSAGE,
+	COST_NETWORK_ERROR,
+	COST_REQUEST_TIMED_OUT,
+	COST_INVALID_SIGNATURE,
+	COST_WRONG_PARA,
+	COST_UNNEEDED_COLLATOR,
+	COST_REPORT_BAD,
+	BENEFIT_NOTIFY_GOOD,
+];
+
+#[derive(Copy, Clone)]
+#[repr(usize)]
+pub enum FitnessEvent {
+	Unexpected = 0,
+	Corrupted = 1,
+	NetworkError = 2,
+	Timeout = 3,
+	SigError = 4,
+	ParaError = 5,
+	Superfluous = 6,
+	ReportBad = 7,
+	NotifyGood = 8,
+}
+
+#[derive(Copy, Clone)]
+pub struct FitnessMetric {
+	last: std::time::Instant,
+	avg: f64,
+	total: u64,
+}
+
+impl PartialEq for FitnessMetric {
+	fn eq(&self, other: &Self) -> bool {
+		(self.avg - other.avg).abs() < f64::EPSILON
+	}
+}
+
+impl Eq for FitnessMetric {}
+
+impl PartialOrd for FitnessMetric {
+	fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Ord for FitnessMetric {
+	fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+		if self.avg < other.avg {
+			std::cmp::Ordering::Less
+		} else {
+			std::cmp::Ordering::Greater
+		}
+	}
+}
+
+impl FitnessMetric {
+	pub fn new() -> Self {
+		Self {
+			last: std::time::Instant::now(),
+			avg: 0f64,
+			total: 0u64,
+		}
+	}
+
+	pub fn insert_event(&mut self) {
+		let mut cumulative = if self.avg == 0f64 {
+			0f64
+		} else {
+			self.total as f64 / self.avg
+		};
+		self.total += 1;
+		cumulative += self.last.elapsed().as_secs() as f64;
+		self.avg = self.total as f64 / cumulative;
```

Comment:
You're right. If the time_elapsed() since last is 0, then this will still be 0.

I'm debating whether it's necessary to keep track of the average independently. Instead we should probably sum the time and the total and then divide when we compute the weighted average.

I will adjust